channels_redis/core.py

```diff
@@ -350,26 +356,38 @@ def _map_channel_to_connection(self, channel_names, message):
         We also return a mapping from channel names to their corresponding Redis
         keys, and a mapping of channels to their capacity
         """
-        connection_to_channels = collections.defaultdict(list)
-        channel_to_message = dict()
-        channel_to_capacity = dict()
-        channel_to_key = dict()
+        channel_keys = list()
+        connection_to_channel_keys = collections.defaultdict(list)
+        channel_key_to_message = dict()
+        channel_key_to_capacity = dict()
 
         for channel in channel_names:
+
             channel_non_local_name = channel
+
             if "!" in channel:
-                message = dict(message.items())
-                message["__asgi_channel__"] = channel
                 channel_non_local_name = self.non_local_name(channel)
+
             channel_key = self.prefix + channel_non_local_name
-            idx = self.consistent_hash(channel_non_local_name)
-            connection_to_channels[idx].append(channel_key)
-            channel_to_capacity[channel] = self.get_capacity(channel)
-            channel_to_message[channel] = self.serialize(message)
-            # We build a
-            channel_to_key[channel] = channel_key
 
-        return connection_to_channels, channel_to_message, channel_to_capacity, channel_to_key
+            if channel_key not in channel_keys:
+                channel_keys.append(channel_key)
+                idx = self.consistent_hash(channel_non_local_name)
+                connection_to_channel_keys[idx].append(channel_key)
+
+            if channel_key not in channel_key_to_message.keys():
+                message = dict(message.items())
+                message["__asgi_channel__"] = [channel]
+                channel_key_to_message[channel_key] = message
+                channel_key_to_capacity[channel_key] = self.get_capacity(channel)
+            else:
+                channel_key_to_message[channel_key]["__asgi_channel__"].append(channel)
+
+        for key in channel_keys:
+            channel_key_to_message[key]["__asgi_channel__"] = ",".join(channel_key_to_message[key]["__asgi_channel__"])
+            channel_key_to_message[key] = self.serialize(channel_key_to_message[key])
+
+        return channel_keys, connection_to_channel_keys, channel_key_to_message, channel_key_to_capacity
```

Comment:
This block needs some serious comments to say what's happening - I'm finding it hard to read.

Reply:
on it